core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -1,126 +1,84 @@
 import org.infinispan.context.InvocationContext;
 import org.infinispan.context.impl.FlagBitSets;
 import org.infinispan.context.impl.TxInvocationContext;
 import org.infinispan.factories.annotations.Inject;
-import org.infinispan.interceptors.InvocationSuccessFunction;
-import org.infinispan.statetransfer.StateTransferManager;
-import org.infinispan.transaction.impl.LocalTransaction;
+import org.infinispan.interceptors.distribution.TxDistributionInterceptor;
 import org.infinispan.util.logging.Log;
 import org.infinispan.util.logging.LogFactory;
 
 /**
- * Locking interceptor to be used by pessimistic caches.
- * Design note: when a lock "k" needs to be acquired (e.g. cache.put("k", "v")), if the lock owner is the local node,
- * no remote call is performed to migrate locking logic to the other (numOwners - 1) lock owners. This is a good
- * optimisation for  in-vm transactions: if the local node crashes before prepare then the replicated lock information
- * would be useless as the tx is rolled back. OTOH for remote hotrod/transactions this additional RPC makes sense because
- * there's no such thing as transaction originator node, so this might become a configuration option when HotRod tx are
- * in place.
- *
- * Implementation note: current implementation acquires locks remotely first and then locally. This is required
- * by the deadlock detection logic, but might not be optimal: acquiring locks locally first might help to fail fast the
- * in the case of keys being locked.
+ * Locking interceptor to be used by pessimistic transactions.
+ * <p>
+ * Design (from 9.1):
+ * <p>
+ * This interceptor is skipped when {@link Flag#SKIP_LOCKING} is used or when all affected keys are already locked.
+ * Read-only commands can acquire the locks when {@link Flag#FORCE_WRITE_LOCK}.
+ * <p>
+ * The new design has 2 concepts. The primary lock and the backup lock. The primary lock is acquired in the primary
+ * owner and only on transaction can acquire at each give time. The backup lock exists in the backup owner and multiple
+ * transactions can acquire simultaneously. The backup lock is associated to the current topology id and the primary
+ * lock can only be acquired if doesn't exists any backup lock with older topology id.
+ * <p>
+ * Since the primary owner doesn't acquire backup locks, they are only used when the primary ownership of a key is
+ * changed.
+ * <p>
+ * The new algorithm will always acquire both. So, lock acquisition (single or multiple keys) will generate a RPC to all
+ * the owners of the key(s). This is triggered in {@link TxDistributionInterceptor#visitLockControlCommand(TxInvocationContext,
+ * LockControlCommand)}.
+ * <p>
+ * The {@link Flag#CACHE_MODE_LOCAL} changes this behaviour. It makes the lock acquisition only local. So, only the
+ * primary or backup lock is acquired. No RPC is create.
+ * <p>
+ * Know Issue: There is small deadlock probably is 2 or more transaction try to acquire the same lock and the primary
```

Comment:
You haven't explain how the backup lock is used (just that it's acquired non-exclusively), so this note is missing a "big picture". Does the to-be-primary node automatically convert its backup lock into primary lock? What if it has the lock locked multiple times?

Reply:
@rvansa @pruivo in an optimistic cache, the backup lock would be converted to a primary lock by replaying the prepare command during commit. Here I assume there is no change, just that an existing backup lock for key `k` prevents any transaction from acquiring a primary lock for `k`. So other transactions cannot lock `k`, but the transaction that acquired the primary lock on the old primary owner can still commit.